modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/components/rule-builder/Calculator/Calculator.es.js

```diff
@@ -0,0 +1,267 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import ClayLayout from '@clayui/layout';
+import {Token, Tokenizer} from 'dynamic-data-mapping-form-builder';
+import {useStateSafe} from 'frontend-js-react-web';
+import React, {forwardRef, useCallback, useEffect, useMemo} from 'react';
+
+import CalculatorButtonArea from './CalculatorButtonArea.es';
+import CalculatorDisplay from './CalculatorDisplay.es';
+import Dropdown from './Dropdown.es';
+
+function getRepeatableFields(fields) {
+	return fields.filter(({repeatable}) => repeatable === true);
+}
+
+function getStateBasedOnExpression(expression) {
+	let disableDot = false;
+	let disableFunctions = false;
+	let disableNumbers = false;
+	let disableOperators = false;
+	let showOnlyRepeatableFields = false;
+	const tokens = Tokenizer.tokenize(expression);
+
+	if (
+		tokens.length > 1 &&
+		tokens[tokens.length - 1].type === Token.LEFT_PARENTHESIS &&
+		tokens[tokens.length - 2].type === Token.FUNCTION &&
+		tokens[tokens.length - 2].value === 'sum'
+	) {
+		disableFunctions = true;
+		disableNumbers = true;
+		disableOperators = true;
+		showOnlyRepeatableFields = true;
+	}
+
+	if (
+		tokens.length === 0 ||
+		(tokens.length > 0 && tokens[tokens.length - 1].type !== Token.LITERAL)
+	) {
+		disableDot = true;
+	}
+
+	if (
+		tokens.length > 0 &&
+		tokens[tokens.length - 1].type === Token.OPERATOR
+	) {
+		disableOperators = true;
+	}
+
+	return {
+		disableDot,
+		disableFunctions,
+		disableNumbers,
+		disableOperators,
+		showOnlyRepeatableFields,
+	};
+}
+
+function shouldAddImplicitMultiplication(tokens, newToken) {
+	const lastToken = tokens[tokens.length - 1];
+
+	return (
+		lastToken !== undefined &&
+		((newToken.type === Token.LEFT_PARENTHESIS &&
+			lastToken.type !== Token.OPERATOR &&
+			lastToken.type !== Token.FUNCTION &&
+			lastToken.type !== Token.LEFT_PARENTHESIS) ||
+			(newToken.type === Token.FUNCTION &&
+				lastToken.type !== Token.OPERATOR &&
+				lastToken.type !== Token.LEFT_PARENTHESIS) ||
+			(newToken.type === Token.VARIABLE &&
+				(lastToken.type === Token.VARIABLE ||
+					lastToken.type === Token.LITERAL)) ||
+			(newToken.type === Token.LITERAL &&
+				(lastToken.type === Token.VARIABLE ||
+					lastToken.type === Token.FUNCTION)))
```

Comment:
Well, it would be good to try to decrease this ternary nesting here 🙂

Reply:
![nossa](https://media.gazetadopovo.com.br/2020/02/07164903/Pai-Nosso-660x372.jpg)

I'll take a look

Reply:
haha 😂, don't worry about it now, it will probably get your head knotted.

Reply:
Ok champ, I'll not touch this.

Reply:
See 825242e

@natocesarrego @matuzalemsteles

Reply:
Amen!